src/native/cipher.rs

```diff
@@ -261,3 +258,305 @@ pub fn decrypt_in<P: PoolParams>(eta:Num<P::Fr>, memo:&[u8], params:&P)->Vec<Opt
         vec![]
     }
 }
+
+/// get encrypted memo chunks with associated decryption keys (chunk: account or note)
+/// returns vector of tupple (index, chunk, key)
+/// indexes are zero-based and enumerated within current memo
+pub fn symcipher_decryption_keys<P: PoolParams>(eta:Num<P::Fr>, mut memo:&[u8], params:&P) -> Option<Vec<(u64, Vec<u8>, Vec<u8>)>> {
+    let num_size = constants::num_size_bits::<P::Fr>()/8;
+    let account_size = constants::account_size_bits::<P::Fr>()/8;
+    let note_size = constants::note_size_bits::<P::Fr>()/8;
+
+    let nozero_items_num = u32::deserialize(&mut memo).ok()? as usize;
+    if nozero_items_num == 0 {
+        return None;
+    }
+
+    let nozero_notes_num = nozero_items_num - 1;
+    let shared_secret_ciphertext_size = nozero_items_num * constants::U256_SIZE + constants::POLY_1305_TAG_SIZE;
+
+    let account_hash = Num::deserialize(&mut memo).ok()?;
+    let note_hashes = buf_take(&mut memo, nozero_notes_num * num_size)?;
+
+    let shared_secret_text = {
+        let a_p = EdwardsPoint::subgroup_decompress(Num::deserialize(&mut memo).ok()?, params.jubjub())?;
+        let ecdh = a_p.mul(eta.to_other_reduced(), params.jubjub());
+        let key = {
+            let mut x: [u8; 32] = [0; 32];
+            ecdh.x.serialize(&mut &mut x[..]).unwrap();
+            keccak256(&x)
+        };
+        let ciphertext = buf_take(&mut memo, shared_secret_ciphertext_size)?;
+        symcipher_decode::<SHARED_SECRETS_HEAPLESS_SIZE>(&key, ciphertext)
+    };
+
+    if let Some(shared_secret_text) = shared_secret_text {
+        // here is a our transaction, we can restore account and all notes
+        let mut shared_secret_text_ptr = shared_secret_text.as_slice();
+
+        let account_key= <[u8;constants::U256_SIZE]>::deserialize(&mut shared_secret_text_ptr).ok()?;
+        let note_key = (0..nozero_notes_num).map(|_| <[u8;constants::U256_SIZE]>::deserialize(&mut shared_secret_text_ptr)).collect::<Result<Vec<_>,_>>().ok()?;
```

Comment:
Let's format this file after the review :) I think it will look much better this way.

Reply:
What kind of formatting do you mean?

Reply:
I meant rustfmt.

Reply:
But it's not mandatory, I think we can format all files later.

Reply:
Okay, it's a good idea. I suggest to plan a great refactor session for core libs. But let's postpone it for the next month, when hackaton will passed. Wdyt?

Reply:
Yes, let's do it this way.